File: src/ListNode.cpp
#include "ListNode.h"
#include "Application.h"

ListNode::ListNode(gc::static_ptr<Application> app):
    _back(nullptr),
    _fwd(nullptr),
    _app(app),
    _id(ListNode::_next_id++)
{
}

void ListNode::addAfter(gc::static_ptr<ListNode> back)
{
    _back = back;
    _fwd = _back->_fwd;

    _back->_fwd = this;
    if (_fwd != nullptr) _fwd->_back = this;
}

void ListNode::unlink()
{
    if (_back != nullptr)
    {
        _back->_fwd = _fwd;
    }
    else
    {
        _app->_start_ptr = _fwd;
    }

    if (_fwd != nullptr) _fwd->_back = _back;
}

File: java_equivalent/ListNode.java
class ListNode
{
    public ListNode back;
    public ListNode fwd;

    public final int id;

    public static int nextId = 0;

    public ListNode()
    {
        id = nextId++;
    }

    public void addAfter(final ListNode back)
    {
        this.back = back;
        this.fwd = this.back.fwd;

        back.fwd = this;
        if (fwd != null) fwd.back = this;
    }

    public void unlink()
    {
        if (back != null)
        {
            back.fwd = fwd;
        }
        else
        {
            App.startPtr = fwd;
        }

        if (fwd != null) fwd.back = back;
    }
}
